community/kernel/src/main/java/org/neo4j/kernel/NeoStoreDataSource.java

```diff
@@ -1094,17 +1094,42 @@ public synchronized void stop()
         {
             return;
         }
-        transactionLogModule.logRotationControl().awaitAllTransactionsClosed();
-        transactionLogModule.logRotationControl().forceEverything();
-        /*
-         * We simply increment the version, essentially "rotating" away
-         * the current active log file, to avoid having a recovery on
-         * next startup. Not necessary, simply speeds up the startup
-         * process.
-         */
-        neoStoreModule.neoStore().incrementAndGetVersion();
 
-        life.shutdown();
+        LogRotationControl logRotationControl = transactionLogModule.logRotationControl();
+
+        // First kindly await all committing transactions to close. Do this without interfering with the
+        // log file monitor. Keep in mind that at this point the availability guard is raised and some time spent
+        // awaiting active transaction to close, on a more coarse-grained level, so no new transactions
+        // should get started. With that said there's actually a race between checking the availability guard
+        // in beginTx, incrementing number of open transactions and raising the guard in shutdown, there might
+        // be some in flight that will get to commit at some point
+        // in the future. Such transactions will fail if they come to commit after our synchronized block below.
+        // Here we're zooming in and focusing on getting committed transactions to close.
+        logRotationControl.awaitAllTransactionsClosed();
+        synchronized ( transactionLogModule.logFile() )
```

Comment:
"lifting up" this object in order to be able to lock it like this makes me a bit uneasy, but not so much that I'll let it stop this fix.